proxy/src/connection.rs

```diff
@@ -162,20 +181,28 @@ impl Future for Connecting {
     type Error = io::Error;
 
     fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
-        let socket = try_ready!(self.0.poll());
+        let socket = try_ready!(self.inner.poll());
         set_nodelay_or_warn(&socket);
-        Ok(Async::Ready(Connection::new(Box::new(socket))))
+        Ok(Async::Ready(Connection::new(socket, self.tls_status)))
     }
 }
 
 // ===== impl Connection =====
 
 impl Connection {
-    /// A constructor of `Connection` with a plain text TCP socket.
-    fn new(io: Box<Io>) -> Self {
+    fn new<I: Io + 'static>(io: I, tls_status: TlsStatus) -> Self {
         Connection {
-            io,
+            io: Box::new(io),
+            peek_buf: BytesMut::new(),
+            tls_status,
```

Comment:
The `Io` trait should be extended to report TLS status. Then we don't need to store TLS status in a separate field in this `Connection` struct and also we don't need to pass in an explicit TLS status in the constructor.

Reply:
Hmm, this is a good idea. 

However, I think that the `Io` trait will be able to report whether or not the connection is over TLS, but not  _why_ the connection is not TLS; it doesn't have a way of knowing whether the connection doesn't have TLS enabled or the TLS config was unavailable (or, at least, no way of knowing that's obvious to me)...

Reply:
> However, I think that the Io trait will be able to report whether or not the connection is over TLS, but not why the connection is not TLS; it doesn't have a way of knowing whether the connection doesn't have TLS enabled or the TLS config was unavailable (or, at least, no way of knowing that's obvious to me)...

Good point. That means only the plaintext constructor needs an explicit flag, but OTOH that would make things more complicated than the current code, so never mind.

Reply:
It was a good suggestion.